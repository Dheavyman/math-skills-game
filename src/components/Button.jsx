import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  selectedNumbers: PropTypes.arrayOf(PropTypes.number).isRequired,
  isAnswerCorrect: PropTypes.bool,
  redraws: PropTypes.number.isRequired,
  countdownRunning: PropTypes.bool.isRequired,
  handleCheckAnswer: PropTypes.func.isRequired,
  handleAcceptAnswer: PropTypes.func.isRequired,
  handleRedraw: PropTypes.func.isRequired,
};

const defaultProps = {
  isAnswerCorrect: null,
};

/**
 * Button component
 *
 * @param {object} props - Properties passed to component
 *
 * @returns {object} React element
 */
const Button = (props) => {
  const {
    selectedNumbers,
    isAnswerCorrect,
    redraws,
    countdownRunning,
    handleCheckAnswer,
    handleAcceptAnswer,
    handleRedraw,
  } = props;
  let button;

  switch (isAnswerCorrect) {
    case true:
      button = (
        <button
          type="button"
          className="btn btn-success btn-lg"
          onClick={handleAcceptAnswer}
        >
          <i className="fa fa-check" />
        </button>
      );
      break;
    case false:
      button = (
        <button
          type="button"
          className="btn btn-danger btn-lg"
        >
          <i className="fa fa-times" />
        </button>
      );
      break;
    default:
      button = (
        <button
          type="button"
          className="btn btn-secondary btn-lg"
          onClick={handleCheckAnswer}
          disabled={!countdownRunning || selectedNumbers.length === 0}
        >
        =
        </button>
      );
  }
  return (
    <div className="col-2 check-btn text-center">
      {button}
      <br />
      <br />
      <button
        type="button"
        className="btn btn-warning btn-sm"
        onClick={handleRedraw}
        disabled={!countdownRunning || redraws <= 0}
      >
        <i className={`fa fa-sync-alt ${redraws > 0 && 'fa-spin'}`} />
        {`  ${redraws}`}
      </button>
    </div>
  );
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
