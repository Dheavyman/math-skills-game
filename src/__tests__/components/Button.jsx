import React from 'react';
import TestRenderer from 'react-test-renderer';

import Button from '../../components/Button';

describe('Button component', () => {
  let props;

  beforeEach(() => {
    props = {
      selectedNumbers: [],
      isAnswerCorrect: null,
      redraws: 5,
      countdownRunning: false,
      handleCheckAnswer: jest.fn(),
      handleAcceptAnswer: jest.fn(),
      handleRedraw: jest.fn()
    };
  });

  it('should render the component with default button successfully', () => {
    const renderer = TestRenderer.create(<Button {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it(`should render the default button when no number has been selected and the
    countdown is running`, () => {
    props = {
      ...props,
      countdownRunning: true
    };
    const renderer = TestRenderer.create(<Button {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render the component with check button when answer is correct',
    () => {
      props = {
        ...props,
        isAnswerCorrect: true,
        countdownRunning: true
      };
      const renderer = TestRenderer.create(<Button {...props} />);

      expect(renderer.toJSON()).toMatchSnapshot();
    });

  it('should render the component with x(fa-times) button when answer is wrong',
    () => {
      props = {
        ...props,
        isAnswerCorrect: false,
        countdownRunning: true
      };
      const renderer = TestRenderer.create(<Button {...props} />);

      expect(renderer.toJSON()).toMatchSnapshot();
    });

  it('should call handleCheckAnswer handler when default button is clicked',
    () => {
      props = {
        ...props,
        isAnswerCorrect: null,
        countdownRunning: true,
        selectedNumbers: [5, 2]
      };
      const renderer = TestRenderer.create(<Button {...props} />);

      renderer.root.findAllByType('button')[0].props.onClick();
      expect(props.handleCheckAnswer).toHaveBeenCalled();
    });

  it('should call handleAcceptAnswer handler when check button is clicked',
    () => {
      props = {
        ...props,
        isAnswerCorrect: true,
        countdownRunning: true,
        selectedNumbers: [5, 2]
      };
      const renderer = TestRenderer.create(<Button {...props} />);

      renderer.root.findAllByType('button')[0].props.onClick();
      expect(props.handleAcceptAnswer).toHaveBeenCalled();
    });

  it('should call handleRedraws handler when redraw button is clicked',
    () => {
      props = {
        ...props,
        countdownRunning: true,
      };
      const renderer = TestRenderer.create(<Button {...props} />);

      renderer.root.findAllByType('button')[1].props.onClick();
      expect(props.handleRedraw).toHaveBeenCalled();
    });
});
