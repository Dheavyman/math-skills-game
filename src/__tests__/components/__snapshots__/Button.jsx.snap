// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render the component with check button when answer is correct 1`] = `
<div
  className="col-2 check-btn text-center"
>
  <button
    className="btn btn-success btn-lg"
    onClick={[MockFunction]}
    type="button"
  >
    <i
      className="fa fa-check"
    />
  </button>
  <br />
  <br />
  <button
    className="btn btn-warning btn-sm"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
  >
    <i
      className="fa fa-sync-alt fa-spin"
    />
      5
  </button>
</div>
`;

exports[`Button component should render the component with default button successfully 1`] = `
<div
  className="col-2 check-btn text-center"
>
  <button
    className="btn btn-secondary btn-lg"
    disabled={true}
    onClick={[MockFunction]}
    type="button"
  >
    =
  </button>
  <br />
  <br />
  <button
    className="btn btn-warning btn-sm"
    disabled={true}
    onClick={[MockFunction]}
    type="button"
  >
    <i
      className="fa fa-sync-alt fa-spin"
    />
      5
  </button>
</div>
`;

exports[`Button component should render the component with x(fa-times) button when answer is wrong 1`] = `
<div
  className="col-2 check-btn text-center"
>
  <button
    className="btn btn-danger btn-lg"
    type="button"
  >
    <i
      className="fa fa-times"
    />
  </button>
  <br />
  <br />
  <button
    className="btn btn-warning btn-sm"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
  >
    <i
      className="fa fa-sync-alt fa-spin"
    />
      5
  </button>
</div>
`;

exports[`Button component should render the default button when no number has been selected and the
    countdown is running 1`] = `
<div
  className="col-2 check-btn text-center"
>
  <button
    className="btn btn-secondary btn-lg"
    disabled={true}
    onClick={[MockFunction]}
    type="button"
  >
    =
  </button>
  <br />
  <br />
  <button
    className="btn btn-warning btn-sm"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
  >
    <i
      className="fa fa-sync-alt fa-spin"
    />
      5
  </button>
</div>
`;
